examples/seismic/tutorials/fwi_gradient_marmousi.py

```diff
@@ -0,0 +1,118 @@
+import argparse
+from memory_profiler import memory_usage
+
+import numpy as np
+from scipy.ndimage import gaussian_filter
+
+from examples.seismic import (
+    Model, AcquisitionGeometry, Receiver, plot_image
+)
+from examples.seismic.acoustic import AcousticWaveSolver
+from devito import configuration, Function, norm, Eq, Operator
+
+configuration['log-level'] = 'WARNING'
+
+# Common Configuration
+nshots = 30  # Number of shots
+nreceivers = 300  # Number of receivers per shot
+
+
+# Define the water layer mask
+def mask(model, value):
+    mask = model > value
+    return mask.astype(int)
+
+
+# Compute residual
+def compute_residual(residual, dobs, dsyn):
+    if residual.grid.distributor.is_parallel:
+        assert np.allclose(dobs.coordinates.data[:], dsyn.coordinates.data)
+        assert np.allclose(residual.coordinates.data[:], dsyn.coordinates.data)
+        diff_eq = Eq(
+            residual, dsyn.subs({dsyn.dimensions[-1]: residual.dimensions[-1]}) -
+            dobs.subs({dobs.dimensions[-1]: residual.dimensions[-1]})
+        )
+        Operator(diff_eq)()
+    else:
+        residual.data[:] = dsyn.data[:] - dobs.data[:]
+    return residual
+
+
+# FWI Gradient Kernel
+def fwi_gradient(mode, model, solver, geometry, source_locations, vp_in, factor=None):
+    grad = Function(name="grad", grid=model.grid)
+    residual = Receiver(name='residual', grid=model.grid,
+                        time_range=geometry.time_axis,
+                        coordinates=geometry.rec_positions)
+    d_obs = Receiver(name='d_obs', grid=model.grid,
+                     time_range=geometry.time_axis,
+                     coordinates=geometry.rec_positions)
+    d_syn = Receiver(name='d_syn', grid=model.grid,
+                     time_range=geometry.time_axis,
+                     coordinates=geometry.rec_positions)
+    objective = 0.0
+    for i in range(nshots):
+        geometry.src_positions[0, :] = source_locations[i, :]
```

Comment:
Modifying geometry isn't good here. Instead get a source `src=geometry.src` and on;y modify the `src` coordinates. (I know this is just taken from the FWI tutorial but it's an old tutorial that needs updates)